File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root == None:return root
        return_li = []
        sub_li = [root.val]
        def dfs(node,target):
            if node.left == None and node.right == None:
                if target == 0:
                    return_li.append(sub_li[:])
                return
            if node.left != None:
                sub_li.append(node.left.val)
                dfs(node.left,target-node.left.val)
                sub_li.pop()
                
            if node.right != None:
                sub_li.append(node.right.val)
                dfs(node.right,target-node.right.val)
                sub_li.pop()
                
        dfs(root,targetSum - root.val)
        return return_li
            
                
                
        

File: 113-path-sum-ii/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> path = new ArrayList<>();
    int targetSum;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        this.targetSum = targetSum;
        dfs(root);
        return ans;
    }
    
    public void dfs(TreeNode node){
        if (node == null)
            return;
        path.add(node.val);
        if (node.left == null && node.right == null){
            int pathsum = 0;
            List<Integer> new_path = new ArrayList<>();
            for (int i : path){
                pathsum += i;
                new_path.add(i);
            }
            if (pathsum == targetSum){
                ans.add(new_path);
            }
        }
        dfs(node.left);
        dfs(node.right);
        path.remove(path.size()-1);
        return;
    }
}